### ISSUE TITLE
iApp deployment within partition with non-default route domain

### ISSUE THREAD

**Issue by colin-stubbs** [f5devcentral/f5-tmsh2iapp#3]
tmsh2iapp's most recent version 20171013.2 (commit 2127b835a937ea88a914745bb820e6c7c8d0f2a2) doesn't handle deployment scenarios within partition with a non-default (non-zero) route domain well.

This is a combination of BIG-IP behaviour and iApp behaviour. If using iWorkflow 2.3.0 with multi-tenancy enabled you WILL run into this issue if trying to use tmsh2iapp to generate iApp templates for import to iWorkflow.

e.g. if deploying within partition TENANT that has non-default route domain ID 123, with a pool member address of 203.0.113.4:80 it will try to create a node named "203.0.113.4", if no node yet exists in the partition the node will be created successfully and BIG-IP will automatically append %123 to it.

If an existing node for 203.0.113.4 has already been created BIG-IP will think of the node as 203.0.113.4%123 and deployment of the iApp will error as it a node named 203.0.113.4%123 with address 203.0.113.4%123 already exists YET the iApp is telling BIG-IP to create a node named 203.0.113.4 with address 203.0.113.4%123 ... the append of %123 by BIG-IP is implicit given the context of being within a partition with non-default route domain.

NOTE: The iApp it creates *will* work in the above situation with non-zero route domains provided the user understands that the need to postfix %N route domain identifiers to pool member addresses.

However; assuming the user will remember to do that, or that a third party system leveraging the API will know to postfix a route domain identifier that it may know nothing about, is dumb.

A better approach to this is to have the generated iApp determine the partition, then determine the default route domain for the partition, and if it is a non-zero route domain postfix %N to the node address.

This can be done with the following the iApp implementation code, the following as an example shell for implementation section within iApp when __pm__pool_name__ is used.

```
puts "Starting iApp $tmsh::app_name.app generated with tmsh2iapp version 20171013.2"

set partition "/[lindex [split [tmsh::pwd] /] 1]"
set partition_name "[lindex [split [tmsh::pwd] /] 1]"

set obj [tmsh::get_config auth partition $partition_name default-route-domain]
set routedomainid [tmsh::get_field_value [lindex $obj 0] default-route-domain]

puts "The iApp is being instantiated in @partition $partition with default RD%$routedomainid"
if { $partition == "/" } { puts "Warning: unexpected behaviour when @partition variable is to \"/\"" }

set cfg { %BREVITY% }
set cfg [string map "@service_folder $tmsh::app_name.app @partition $partition %BREVITY% __app_service__ $tmsh::app_name.app/$tmsh::app_name " $cfg]

set fileId [open /var/tmp/cslab_salt_stack_master_v1.0.13.cfg "w"]
puts -nonewline $fileId $cfg
close $fileId

tmsh::load sys config merge file /var/tmp/cslab_salt_stack_master_v1.0.13.cfg

if {([info exists {::pm__%POOL_NAME%}]) && ([string length ${::pm__%POOL_NAME%}] > 0)} {
  if {$routedomainid != 0} {
    set cmd "tmsh::modify ltm pool %POOL_NAME% { members replace-all-with { [string map ": %${routedomainid}:" ${::pm__%POOL_NAME%}] } }"
  } else {
    set cmd "tmsh::modify ltm pool %POOL_NAME% { members replace-all-with { ${::pm__%POOL_NAME%} } }"
  }
  puts "$cmd"
  eval $cmd
  if {([info exists {::pm__%POOL_NAME%_properties}]) && ([string length ${::pm__%POOL_NAME%_properties}] > 0)} {
    set cmd "tmsh::modify ltm pool %POOL_NAME% { members modify { all { ${::pm__%POOL_NAME%_properties} } } }"
    puts "$cmd"
    eval $cmd
  }
}

puts "Finished iApp $tmsh::app_name.app generated with tmsh2iapp version 20171013.2"
```

A helper function that may also be useful, which I've taken from the application services integration iApp, may also be useful to insert for other situations in which tmsh2iapp has to iterate through IP addresses to determine if it should append %N route domain identifiers or not.

```
# Check to see if an ip has a routedomain included.
# Return: 0=false; 1=true
proc has_routedomain { ip } {
  return [string match *%* $ip]
}
```

**Comment by alonsocamaro** [comment]
Many thanks Colins

I understand that the existing node is /Common/blahblah and new node is /MyPartition/blahblah yet the BIG-IP fails to load the configuration. 

I believe this is something that should be fixed in the BIG-IP code but I'm more than happy of handling this misbehavior in tmsh2iapp.

Given that you already have the fix. Allow me some time to reproduce it and fix it and I will take care of creating a support case.

**Comment by colin-stubbs** [comment]
Hi Ulises,

Actually it's more an issue to do with the auto-generated node name... e.g. 
ltm node 203.0.113.0 { address 203.0.113.0%1234 }
vs
ltm node 203.0.113.0%1234 { address 203.0.113.0%1234 }

Which should be considered to be the same thing by TMOS but isn't.

When nodes are auto-created by TMOS as part of pool creation; the *NAME* of the node is derived from the name provided in the pool. 

The %N route domain ID can be explicitly provided, in which case the node name will wind up including it, or it can be implicit based on context for pool creation, in which case the node name will not include %N BUT the node address will be N.N.N.N%1234 in all internal TMOS representation... depending on context TMOS may or may not display %1234 on the node address though.

If the fix was going to be in BIG-IP/TMOS... then default node creation should actually create ltm node 203.0.113.0%1234 { address 203.0.113.0%1234 } **NOT** ltm node 203.0.113.0 { address 203.0.113.0%1234 }

As that is more precise and accurately describes the node uniquely with a unique name.

iWorkflow with the Application Services Integration iApp works that way.

Maybe this is a better example of the issue?

root@(bigip03)(cfg-sync Sync Failed)(Eval:Active)(/)(tmos)# list /ltm recursive node
ltm node CSLAB/192.0.2.8%67 {
    address 192.0.2.8%67
    partition CSLAB
}
ltm node CSLAB/192.0.2.9%67 {
    address 192.0.2.9%67
    partition CSLAB
}
ltm node CSLAB/203.0.113.2 {
    address 203.0.113.2%67
    partition CSLAB
}
ltm node CSLAB/203.0.113.3 {
    address 203.0.113.3%67
    partition CSLAB
}
root@(bigip03)(cfg-sync Sync Failed)(Eval:Active)(/)(tmos)# cd /TEST
root@(bigip03)(cfg-sync Sync Failed)(Eval:Active)(/TEST)(tmos)# list auth partition TEST all-properties
auth partition TEST {
    default-route-domain 1234
    description none
}
root@(bigip03)(cfg-sync Sync Failed)(Eval:Active)(/TEST)(tmos)# create ltm pool POOL-Creates-Nodes-1 members replace-all-with { 192.0.2.8:80 } root@(bigip03)(cfg-sync Sync Failed)(Eval:Active)(/TEST)(tmos)# create ltm pool POOL-Creates-Nodes-2 members replace-all-with { 192.0.2.9%1234:80 }
root@(bigip03)(cfg-sync Sync Failed)(Eval:Active)(/TEST)(tmos)# list ltm node
ltm node 192.0.2.8 {
    address 192.0.2.8
    partition TEST
}
ltm node 192.0.2.9%1234 {
    address 192.0.2.9
    partition TEST
}
root@(bigip03)(cfg-sync Sync Failed)(Eval:Active)(/TEST)(tmos)# create ltm pool POOL-Creates-Nodes-3 members replace-all-with { 192.0.2.8:80 192.0.2.9:80 }
0107003a:3: Pool member node (/TEST/192.0.2.9) and existing node (/TEST/192.0.2.9%1234) cannot use the same IP Address (192.0.2.9%1234).
root@(bigip03)(cfg-sync Sync Failed)(Eval:Active)(/TEST)(tmos)# create ltm pool POOL-Creates-Nodes-3 members replace-all-with { 192.0.2.8:80 192.0.2.9%1234:80 }
root@(bigip03)(cfg-sync Sync Failed)(Eval:Active)(/TEST)(tmos)# list ltm pool
ltm pool POOL-Creates-Nodes-1 {
    members {
        192.0.2.8:http {
            address 192.0.2.8
        }
    }
    partition TEST
}
ltm pool POOL-Creates-Nodes-2 {
    members {
        192.0.2.9%1234:http {
            address 192.0.2.9
        }
    }
    partition TEST
}
ltm pool POOL-Creates-Nodes-3 {
    members {
        192.0.2.8:http {
            address 192.0.2.8
        }
        192.0.2.9%1234:http {
            address 192.0.2.9
        }
    }
    partition TEST
}
root@(bigip03)(cfg-sync Sync Failed)(Eval:Active)(/TEST)(tmos)# cd /
root@(bigip03)(cfg-sync Sync Failed)(Eval:Active)(/)(tmos)# list /ltm recursive node
ltm node CSLAB/192.0.2.8%67 {
    address 192.0.2.8%67
    partition CSLAB
}
ltm node CSLAB/192.0.2.9%67 {
    address 192.0.2.9%67
    partition CSLAB
}
ltm node CSLAB/203.0.113.2 {
    address 203.0.113.2%67
    partition CSLAB
}
ltm node CSLAB/203.0.113.3 {
    address 203.0.113.3%67
    partition CSLAB
}
ltm node TEST/192.0.2.8 {
    address 192.0.2.8%1234
    partition TEST
}
ltm node TEST/192.0.2.9%1234 {
    address 192.0.2.9%1234
    partition TEST
}
root@(bigip03)(cfg-sync Sync Failed)(Eval:Active)(/)(tmos)# list /ltm recursive pool
ltm pool CSLAB/POOL-IPv4-ECMP-Default-Gateway {
    members {
        CSLAB/203.0.113.2:any {
            address 203.0.113.2%67
            session monitor-enabled
            state up
        }
        CSLAB/203.0.113.3:any {
            address 203.0.113.3%67
            session monitor-enabled
            state up
        }
    }
    monitor Common/gateway_icmp
    partition CSLAB
}
ltm pool CSLAB/Salt-Master-1.app/default_pool {
    app-service /CSLAB/Salt-Master-1.app/Salt-Master-1
    load-balancing-mode least-connections-member
    members {
        CSLAB/192.0.2.8%67:any {
            address 192.0.2.8%67
            app-service /CSLAB/Salt-Master-1.app/Salt-Master-1
            session monitor-enabled
            state up
        }
        CSLAB/192.0.2.9%67:any {
            address 192.0.2.9%67
            app-service /CSLAB/Salt-Master-1.app/Salt-Master-1
            session monitor-enabled
            state up
        }
    }
    monitor CSLAB/Salt-Master-1.app/MONITOR-Salt-Stack-TCP-4505
    partition CSLAB
}
ltm pool CSLAB/Web-HTTPS-1.app/Web-HTTPS-1_pool_0 {
    app-service /CSLAB/Web-HTTPS-1.app/Web-HTTPS-1
    load-balancing-mode least-connections-member
    members {
        CSLAB/192.0.2.8%67:https {
            address 192.0.2.8%67
            app-service /CSLAB/Web-HTTPS-1.app/Web-HTTPS-1
            session monitor-enabled
            state down
        }
        CSLAB/192.0.2.9%67:https {
            address 192.0.2.9%67
            app-service /CSLAB/Web-HTTPS-1.app/Web-HTTPS-1
            session monitor-enabled
            state up
        }
    }
    monitor CSLAB/Web-HTTPS-1.app/monitor_0
    partition CSLAB
}
ltm pool CSLAB/Web-HTTPS-Offload-1.app/Web-HTTPS-Offload-1_pool_0 {
    app-service /CSLAB/Web-HTTPS-Offload-1.app/Web-HTTPS-Offload-1
    members {
        CSLAB/192.0.2.8%67:http {
            address 192.0.2.8%67
            app-service /CSLAB/Web-HTTPS-Offload-1.app/Web-HTTPS-Offload-1
            session monitor-enabled
            state down
        }
        CSLAB/192.0.2.9%67:http {
            address 192.0.2.9%67
            app-service /CSLAB/Web-HTTPS-Offload-1.app/Web-HTTPS-Offload-1
            session monitor-enabled
            state up
        }
    }
    monitor Common/http
    partition CSLAB
}
ltm pool TEST/POOL-Creates-Nodes-1 {
    members {
        TEST/192.0.2.8:http {
            address 192.0.2.8%1234
        }
    }
    partition TEST
}
ltm pool TEST/POOL-Creates-Nodes-2 {
    members {
        TEST/192.0.2.9%1234:http {
            address 192.0.2.9%1234
        }
    }
    partition TEST
}
ltm pool TEST/POOL-Creates-Nodes-3 {
    members {
        TEST/192.0.2.8:http {
            address 192.0.2.8%1234
        }
        TEST/192.0.2.9%1234:http {
            address 192.0.2.9%1234
        }
    }
    partition TEST
}
root@(bigip03)(cfg-sync Sync Failed)(Eval:Active)(/)(tmos)#

**Comment by alonsocamaro** [comment]
Thanks Colin, 

just to put the wording properly. It's not the node's name but the address automatically generated when not specifying the address of a node, ie: it's possible to have the same node name but in different partitions as long as the address in this case. I believe that BIG-IP's automatically generated address should take care of the user's default routing domain as you suggest.

```
root@(atom-bigip)(cfg-sync Standalone)(Active)(/Common)(tmos)# create ltm node /test/something address 1.2.3.4
root@(atom-bigip)(cfg-sync Standalone)(Active)(/Common)(tmos)# create ltm node /Common/something address 1.2.3.5
```

**Comment by colin-stubbs** [comment]
Sure, that's a work around, though it's not how tmsh2iapp document currently suggests that it should be used.

**Comment by colin-stubbs** [comment]
Pull request now has two commits; second one adds @routedomainid user insertable variable. They can choose to use that to insert automatically detected %RD when relevant.

**Comment by alonsocamaro** [comment]
I wanted to clarify one thing that is not well documented:

- When using as pool member an IP address rather than a arbitrary name that doesn't look like an IP address the IP address must explicitly contain the routing domain. I've researched this and I found this is the expected behaviour in BIG-IP which I agree is not customer friendly.

Your patch (which has just been merged) solves this transparently to the user and in turns also avoids problems like one you mentioned when doing a config-sync.

Many thanks Colin and I'm very sorry for the delay merging your contribution.
